In [2]:
import json
my_list = []
with open('project.json', encoding='utf-8') as f:
    file = json.load(f)
    for smth in file['messages']:
        if smth['text'] != '' and type(smth['text']) == str:
            my_list.append(smth['text'])
        
print(len(my_list))

2630


In [3]:
best_list = []
with open('best_aneks.json', encoding='utf-8') as f:
    file = json.load(f)
    for smth in file['messages']:
        if smth['text'] != '' and type(smth['text']) == str:
            best_list.append(smth['text'])
        
print(len(best_list))

115


In [4]:
crazy_list = []
with open('crazy_aneks.json', encoding='utf-8') as f:
    file = json.load(f)
    for smth in file['messages']:
        if smth['text'] != '' and type(smth['text']) == str:
            crazy_list.append(smth['text'])
        
print(len(crazy_list))

682


In [5]:
newaneks_list = []
with open('newaneks.json', encoding='utf-8') as f:
    file = json.load(f)
    for smth in file['messages']:
        if smth['text'] != '' and type(smth['text']) == str:
            newaneks_list.append(smth['text'])
        
print(len(newaneks_list))

1684


я тут скачала датабазы анеков из тг и потом добавлю их одну большую дб

In [6]:
import sqlite3
from bs4 import BeautifulSoup
import re
import requests
from pprint import pprint

In [7]:
session = requests.session()

In [8]:
from fake_useragent import UserAgent
ua = UserAgent(verify_ssl=False)

установила всякие штуки для краулинга по сайтам: UserAgent, чтобы сайт не лег, суп и requests, чтобы скачивать информацию со страниц сайта

In [9]:
aneksb = []
for page_number in range(1, 1143):
    url = f'https://baneks.ru/{page_number}'
    req = session.get(url, headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    aneksb.append(soup.find('p').text)
print(aneksb)

['— Скажите, а это ваш "Ягуар" стоит около выхода?\n— Да.\n— Я допью?', 'На распродаже человеческих органов нaчалась давка и я еле успел унести нoги.', '— Вы любите Кафку?\n— Да, грефневую!', 'Марк Цукерберг и Павел Дуров заходят в бар.\nЦукерберг: Мне пиво и орешки. \nДуров: Мне то же самое.', 'Скоро выйдет моя книга... хотя зря я ее, конечно, съел...', '— Милая, быстрее сюда! Посмотри, как мороз разукрасил окна.\n— О, господи! Это же твоей рукой написано "ты жирная шлюха"!\n— Нет. Это мороз.', 'Судья на заседании так долго призывал к порядку, что сколотил скворечник.', 'Боксёры не занимаются сексом перед боем. Потому что недолюбливают друг друга.', 'Пожар в Техасе, в школе:\n- Кидай мне детей, я их буду ловить! - крикнул один пожарный другому.\n- Ок.\n- Ты что негров не ловишь?\n- Блять ... А я думал, нахуя ты мне горелых кидаешь?!', 'С успехом Стаса Михайлова и Елены Ваенги все забыли про певца Юлиана. "А ведь я начал петь хуйню гораздо раньше!" — жалуется исполнитель.', 'Полярник И

In [10]:
print(len(aneksb))

1142


In [11]:
url = f'https://ha-ha.mirtesen.ru/blog/43621057282/Anekdotyi-kategorii-B'
req = session.get(url, headers={'User-Agent': ua.random})
page = req.text
soup = BeautifulSoup(page, 'html.parser')
textaneka = soup.find('div', attrs={'class':'blog-post__text','itemprop':'articleBody'}).text
textaneka = textaneka.split('~')
print(len(textaneka))

27


поскачивала анекдотов с сайтов

In [12]:
sobranie_anekov = my_list + aneksb + best_list + crazy_list + newaneks_list + textaneka
s = len(sobranie_anekov)
print(s)

6280


сделала из всех скачанных и накраулеренных анеков список

In [13]:
import pandas as pd 

data = {'anek': sobranie_anekov} 
 
df = pd.DataFrame(data) 
 
print(df)

                                                   anek
0     Выбирает зоофил собаку:\n-Вам какой породы?\n-...
1     петля» – подумал Штирлиц. В дверь постучали. «...
2     В кружке талантов дети пели песню разным стран...
3     едут батя с сыном на шестёрке, перевернулись -...
4     Родительское собрание в школе. Учитель говорит...
...                                                 ...
6275   Один мальчик питался только бутербродами и ум...
6276   Муж собирается в бар. Жена ему говорит: сейча...
6277   Идет мужик по полю, смотрит — колодец. Заглян...
6278   Встречаются весной, после зимовки, Лиса, Волк...
6279   Тихая Украинская ночь, поют цикады... Мужик п...

[6280 rows x 1 columns]


сделала из списка дб, или по-другому будущую контрольную выборку для определения темы анека

In [14]:
vovaaneks = []
for i in range(79):
    url = f'https://allanecdots.ru/vova/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        vovaaneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez = dict.fromkeys(vovaaneks, 'vova')
print(rez)

{'Вовочке на день рождения тетя подарила игрушечный пистолет, стреляющий пластмассовыми патронами. Мама говорит Вовочке:\r\n- Сынок, что нужно сказать тете Клаве? \r\n- Волына чистая? По мокрухе не мелькала?': 'vova', 'М.И. - Дети, расскажите, кто Ваш папа.\r\nМаша. - Мой папа менеджер. \r\nМ.И. - Хорошо. А еще?\r\nВовочка. - А мой папа - хакер!\r\nМ.И. - ?\r\nВовочка: - Он три раза от пьянства кодировался и три раза код ломал.': 'vova', 'Учительница спрашивает в классе:\r\n- Чем отличается жизнь до революции от той, что сейчас? \r\nВовочка:\r\n- До революции говорили: "Пшел вон!" А теперь говорят: "Приходите завтра!"': 'vova', 'Учительница на уроке:\r\n- Дети назовите слово на букву. \r\nИ поглядев с опаской на Вовочку договорила:\r\n- На букву Х, в котором нет буквы У.\r\nВовочка.\r\n- Xeр!': 'vova', 'МарьИвановна говорит Вовочке:\r\n- Я тебе дам презерватив, передай его, пожалуйста, своим родителям и скажи, чтобы больше так не делали. ': 'vova', 'Почтальон приносит заказное письмо в

In [15]:
poshlieaneks = []
for i in range(324):
    url = f'https://allanecdots.ru/adults/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        poshlieaneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(poshlieaneks, 'adults'))

In [16]:
computersaneks = []
for i in range(74):
    url = f'https://allanecdots.ru/computers/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        computersaneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(computersaneks, 'computers'))
print(rez)

{'Вовочке на день рождения тетя подарила игрушечный пистолет, стреляющий пластмассовыми патронами. Мама говорит Вовочке:\r\n- Сынок, что нужно сказать тете Клаве? \r\n- Волына чистая? По мокрухе не мелькала?': 'vova', 'М.И. - Дети, расскажите, кто Ваш папа.\r\nМаша. - Мой папа менеджер. \r\nМ.И. - Хорошо. А еще?\r\nВовочка. - А мой папа - хакер!\r\nМ.И. - ?\r\nВовочка: - Он три раза от пьянства кодировался и три раза код ломал.': 'vova', 'Учительница спрашивает в классе:\r\n- Чем отличается жизнь до революции от той, что сейчас? \r\nВовочка:\r\n- До революции говорили: "Пшел вон!" А теперь говорят: "Приходите завтра!"': 'vova', 'Учительница на уроке:\r\n- Дети назовите слово на букву. \r\nИ поглядев с опаской на Вовочку договорила:\r\n- На букву Х, в котором нет буквы У.\r\nВовочка.\r\n- Xeр!': 'vova', 'МарьИвановна говорит Вовочке:\r\n- Я тебе дам презерватив, передай его, пожалуйста, своим родителям и скажи, чтобы больше так не делали. ': 'vova', 'Почтальон приносит заказное письмо в

In [17]:
blondesaneks = []
for i in range(44):
    url = f'https://allanecdots.ru/blondes/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        blondesaneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(blondesaneks, 'blondes'))

In [18]:
piter_aneks = []
for i in range(9):
    url = f'https://allanecdots.ru/piter/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        piter_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(piter_aneks, 'piter'))

In [19]:
black_aneks = []
for i in range(194):
    url = f'https://allanecdots.ru/black/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        black_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(black_aneks, 'black'))

In [20]:
kinder_aneks = []
for i in range(87):
    url = f'https://allanecdots.ru/kinder/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        kinder_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(kinder_aneks, 'kinder'))

In [21]:
gabrovska_aneks = []
for i in range(3):
    url = f'https://allanecdots.ru/gabrovska/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        gabrovska_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(gabrovska_aneks, 'gabrovska'))

In [22]:
medical_aneks = []
for i in range(150):
    url = f'https://allanecdots.ru/medical/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        medical_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(medical_aneks, 'medical'))

In [23]:
school_student_aneks = []
for i in range(76):
    url = f'https://allanecdots.ru/scool/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        school_student_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(school_student_aneks, 'school_student'))

In [24]:
alcohol_aneks = []
for i in range(143):
    url = f'https://allanecdots.ru/alcohol/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        alcohol_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(alcohol_aneks, 'alcohol'))

In [25]:
political_aneks = []
for i in range(131):
    url = f'https://allanecdots.ru/political/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        political_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(political_aneks, 'political'))

In [26]:
auto_aneks = []
for i in range(58):
    url = f'https://allanecdots.ru/auto/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        auto_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(auto_aneks, 'auto'))

In [27]:
for_child_aneks = []
for i in range(17):
    url = f'https://allanecdots.ru/child/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        for_child_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(for_child_aneks, 'for_child'))

In [28]:
russians_aneks = []
for i in range(50):
    url = f'https://allanecdots.ru/russians/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        russians_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(russians_aneks, 'russians'))

In [29]:
jews_aneks = []
for i in range(83):
    url = f'https://allanecdots.ru/jews/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        jews_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(jews_aneks, 'jews'))

In [30]:
stirlitz_aneks = []
for i in range(11):
    url = f'https://allanecdots.ru/stirlitz/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        stirlitz_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(stirlitz_aneks, 'stirlitz'))

In [31]:
different_aneks = []
for i in range(144):
    url = f'https://allanecdots.ru/different/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        different_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(different_aneks, 'different'))

In [32]:
sport_aneks = []
for i in range(28):
    url = f'https://allanecdots.ru/spotr/{i}/'
    req = session.get(url,  headers={'User-Agent': ua.random})
    page = req.text
    soup = BeautifulSoup(page, 'html.parser')   
    textaneka = soup.find('div', attrs={'class':"container"})
    textaneka2 = textaneka.find_all('p', attrs={'class':"anekdot_body"})
    for smth in textaneka2:
        sport_aneks.append(re.sub(r'allanecdots.ru', r'', smth.text))
rez.update(dict.fromkeys(sport_aneks, 'sport'))

In [33]:
 themed_df = pd.DataFrame(rez.items(), columns=['anek', 'theme'])

In [34]:
themed_df

anek  theme
0      Вовочке на день рождения тетя подарила игрушеч...   vova
1      М.И. - Дети, расскажите, кто Ваш папа.\r\nМаша...   vova
2      Учительница спрашивает в классе:\r\n- Чем отли...   vova
3      Учительница на уроке:\r\n- Дети назовите слово...   vova
4      МарьИвановна говорит Вовочке:\r\n- Я тебе дам ...   vova
...                                                  ...    ...
18080  На лондонском стадионе туман. Судья наталкивае...  sport
18081  Московский "Спартак" продал своего талантливог...  sport
18082  На спортивных состязаниях комментатор:\n- Вним...  sport
18083  Олимпиада. В финал выходят американец и русски...  sport
18084  На соревнованиях по фехтованию у кенийского сп...  sport

[18085 rows x 2 columns]

а тут я очень по-умному 19 раз одинаковым способом скачивала анекддоты с разных страниц по темам в 3 ночи... потом собрала их в дб с темами - будущую обучающую выборку

In [35]:
themed_df["theme"] = themed_df["theme"].map({'vova':1, 'adults':2, 'computers':3, 'blondes':4, 'piter':5, 'black':6, 'kinder':7, 'gabrovska':8, 'medical':9, 'school_student':10, 'alcohol':11, 'political':12, 'auto':13, 'for_child':14, 'russians':15, 'jews':16, 'stirlitz':17, 'different':18, 'sport':19})

In [36]:
themed_df

anek  theme
0      Вовочке на день рождения тетя подарила игрушеч...      1
1      М.И. - Дети, расскажите, кто Ваш папа.\r\nМаша...      1
2      Учительница спрашивает в классе:\r\n- Чем отли...      1
3      Учительница на уроке:\r\n- Дети назовите слово...      1
4      МарьИвановна говорит Вовочке:\r\n- Я тебе дам ...      1
...                                                  ...    ...
18080  На лондонском стадионе туман. Судья наталкивае...     19
18081  Московский "Спартак" продал своего талантливог...     19
18082  На спортивных состязаниях комментатор:\n- Вним...     19
18083  Олимпиада. В финал выходят американец и русски...     19
18084  На соревнованиях по фехтованию у кенийского сп...     19

[18085 rows x 2 columns]

поскольку нам нужно работать с числовыми данными при предсказании заменю каждую тему на ее числовой эквивалент

In [37]:
import nltk
from nltk.tokenize import word_tokenize

In [38]:
import spacy

nlp = spacy.load("en_core_web_sm")

In [39]:
from nltk.corpus import stopwords
sw = stopwords.words('russian')
print(sw)

['и', 'в', 'во', 'не', 'что', 'он', 'на', 'я', 'с', 'со', 'как', 'а', 'то', 'все', 'она', 'так', 'его', 'но', 'да', 'ты', 'к', 'у', 'же', 'вы', 'за', 'бы', 'по', 'только', 'ее', 'мне', 'было', 'вот', 'от', 'меня', 'еще', 'нет', 'о', 'из', 'ему', 'теперь', 'когда', 'даже', 'ну', 'вдруг', 'ли', 'если', 'уже', 'или', 'ни', 'быть', 'был', 'него', 'до', 'вас', 'нибудь', 'опять', 'уж', 'вам', 'ведь', 'там', 'потом', 'себя', 'ничего', 'ей', 'может', 'они', 'тут', 'где', 'есть', 'надо', 'ней', 'для', 'мы', 'тебя', 'их', 'чем', 'была', 'сам', 'чтоб', 'без', 'будто', 'чего', 'раз', 'тоже', 'себе', 'под', 'будет', 'ж', 'тогда', 'кто', 'этот', 'того', 'потому', 'этого', 'какой', 'совсем', 'ним', 'здесь', 'этом', 'один', 'почти', 'мой', 'тем', 'чтобы', 'нее', 'сейчас', 'были', 'куда', 'зачем', 'всех', 'никогда', 'можно', 'при', 'наконец', 'два', 'об', 'другой', 'хоть', 'после', 'над', 'больше', 'тот', 'через', 'эти', 'нас', 'про', 'всего', 'них', 'какая', 'много', 'разве', 'три', 'эту', 'моя', 'впр

In [40]:
nltk.download('wordnet')
from nltk import WordNetLemmatizer

wnl = WordNetLemmatizer()

[nltk_data] Downloading package wordnet to
[nltk_data]     C:\Users\kuzne\AppData\Roaming\nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [41]:
lemmas = []
for smth in themed_df['anek']:
    words = [w.lower() for w in word_tokenize(smth) if w.isalpha()]
    words_new = ' '.join(words)
    words_new_new = nlp(words_new)
    lemmas_small = []
    for i in words_new_new:
        i1 = i.lemma_
        lemmas_small.append(i1)
    lemmas.append(' '.join(lemmas_small))
themed_df['parsed_aneks'] = lemmas

In [42]:
themed_df.head()

anek  theme  \
0  Вовочке на день рождения тетя подарила игрушеч...      1   
1  М.И. - Дети, расскажите, кто Ваш папа.\r\nМаша...      1   
2  Учительница спрашивает в классе:\r\n- Чем отли...      1   
3  Учительница на уроке:\r\n- Дети назовите слово...      1   
4  МарьИвановна говорит Вовочке:\r\n- Я тебе дам ...      1   

                                        parsed_aneks  
0  вовочке на день рождения тетя подарила игрушеч...  
1  дети расскажите кто ваш папа маша мой папа мен...  
2  учительница спрашивает в классе чем отличается...  
3  учительница на уроке дети назовите слово на бу...  
4  марьивановна говорит вовочке я тебе дам презер...

тут я парсила анеки с помощью нлтк и спэйси, сделала с ними отдельный столбец в обучающей дб, снизу делаю в контрольной

In [43]:
lemmas = []
for smth in df['anek']:
    words = [w.lower() for w in word_tokenize(smth) if w.isalpha()]
    words_new = ' '.join(words)
    words_new_new = nlp(words_new)
    lemmas_small = []
    for i in words_new_new:
        i1 = i.lemma_
        lemmas_small.append(i1)
    lemmas.append(' '.join(lemmas_small))
df['parsed_aneks'] = lemmas

In [44]:
df.head()

anek  \
0  Выбирает зоофил собаку:\n-Вам какой породы?\n-...   
1  петля» – подумал Штирлиц. В дверь постучали. «...   
2  В кружке талантов дети пели песню разным стран...   
3  едут батя с сыном на шестёрке, перевернулись -...   
4  Родительское собрание в школе. Учитель говорит...   

                                        parsed_aneks  
0        выбирает зоофил собаку какой породы поебать  
1  петля подумал штирлиц в дверь постучали временная  
2  в кружке талантов дети пели песню разным стран...  
3  едут батя с сыном на шестёрке перевернулись ед...  
4  родительское собрание в школе учитель говорит ...

In [45]:
from sklearn.feature_extraction.text import CountVectorizer
cvec = CountVectorizer(binary=True, max_features=10000, stop_words=sw)
cvec.fit(themed_df['parsed_aneks'])

CountVectorizer(binary=True, max_features=10000,
                stop_words=['и', 'в', 'во', 'не', 'что', 'он', 'на', 'я', 'с',
                            'со', 'как', 'а', 'то', 'все', 'она', 'так', 'его',
                            'но', 'да', 'ты', 'к', 'у', 'же', 'вы', 'за', 'бы',
                            'по', 'только', 'ее', 'мне', ...])

In [46]:
X_train = cvec.transform(themed_df['parsed_aneks']).astype(int)
X_test = cvec.transform(df['parsed_aneks']).astype(int)

тут я превратила все анеки в вектора, чтобы потом делать предсказание

In [47]:
from sklearn.svm import LinearSVC
model = LinearSVC(C=0.01)

я выбрала модель LinearSVC (использовала ее как-то в дз, понравился результат)

In [48]:
model.fit(X_train, themed_df['theme'])

LinearSVC(C=0.01)

обучаю модель

In [49]:
y_test = model.predict(X_test)

In [50]:
df['theme'] = y_test

здесь я записываю получившиеся пресказания в дб

In [51]:
df.head(30)

anek  \
0   Выбирает зоофил собаку:\n-Вам какой породы?\n-...   
1   петля» – подумал Штирлиц. В дверь постучали. «...   
2   В кружке талантов дети пели песню разным стран...   
3   едут батя с сыном на шестёрке, перевернулись -...   
4   Родительское собрание в школе. Учитель говорит...   
5           На украинском торренте раздается хрюканье   
6   Дагестанские блохи услышав лезгинку затоптали ...   
7   дерутся русский и беременная, всё по честному,...   
8   -Саш, а что тебе нравится в девушках?\n-ну как...   
9   Был один парень, которого заколдовал колдун и ...   
10  Все мечтали работать у Стива Джобса. Кроме его...   
11  Жена просит мужа:\n-слушай, сыну уже 15, погов...   
12  Заходят два дракона в бар.\nОдин другого спраш...   
13  Идет заяц по лесу, навстречу ему медведь.\n- З...   
14  Приходят в роддом негр, русский и хохол. Выход...   
15  В дверь постучали 8 раз.\n"Осьминог", – подума...   
16  - пошли ко мне, у меня чайный гриб есть\n- нет...   
17  В автобусе едет старушка, рядом стоит пионер. ...   
18  Садится Штирлиц в такси и говорит:\n- Трогай!\...   
19  Поручик Ржевский спрашивает у корнета: -Корнет...   
20  Заблудилась бабулька, хочет у кого-нибудь доро...   
21  Летят на самолете англичанин, француз, немка и...   
22  Знаете, почему есть сигареты Пётр I, но нет си...   
23  У мужика вздулся живот. Он к врачу. А врач реш...   
24  Молдаване после секса отворачиваются к стенке ...   
25  В почти пустой автобус заходит хиппи и, увидев...   
26  — Поручик, вы охуели! Я вызываю вас на дуэль!\...   
27  У маленького мальчика мама забеременела. В оди...   
28  У Сталина было два танка.Он катался на них по ...   
29  Видит араб, как человек выбивает ковёр, он к н...   

                                         parsed_aneks  theme  
0         выбирает зоофил собаку какой породы поебать      6  
1   петля подумал штирлиц в дверь постучали временная     17  
2   в кружке талантов дети пели песню разным стран...     12  
3   едут батя с сыном на шестёрке перевернулись ед...      2  
4   родительское собрание в школе учитель говорит ...      1  
5           на украинском торренте раздается хрюканье      2  
6   дагестанские блохи услышав лезгинку затоптали ...      6  
7   дерутся русский и беременная всё по честному б...     15  
8   а что тебе нравится в девушках как грудь и поп...      2  
9   был один парень которого заколдовал колдун и о...      2  
10  все мечтали работать у стива джобса кроме его ...     18  
11  жена просит мужа сыну уже поговори с ним о пол...      2  
12  заходят два дракона в бар один другого спрашив...      6  
13  идет заяц по лесу навстречу ему медведь заяц д...      2  
14  приходят в роддом негр русский и хохол выходит...     15  
15     в дверь постучали раз осьминог подумал штирлиц     17  
16  пошли ко мне у меня чайный гриб есть нет я нас...      2  
17  в автобусе едет старушка рядом стоит пионер ст...      7  
18  садится штирлиц в такси и говорит трогай такси...     17  
19  поручик ржевский спрашивает у корнета вы имеет...      2  
20  заблудилась бабулька хочет у дорогу спросить а...      6  
21  летят на самолете англичанин француз немка и р...      2  
22  знаете почему есть сигареты пётр I но нет сига...      6  
23  у мужика вздулся живот он к врачу а врач решил...      9  
24  молдаване после секса отворачиваются к стенке ...      2  
25  в почти пустой автобус заходит хиппи и увидев ...      2  
26  поручик вы охуели я вызываю вас на дуэль выбир...      2  
27  у маленького мальчика мама забеременела в один...      7  
28  у сталина было два катался на них по очереди о...      9  
29  видит араб как человек выбивает ковёр он к нем...      6

In [52]:
df['theme'].value_counts()

2     2710
6     1035
18     507
11     337
7      303
12     270
9      267
10     211
16     163
13     111
17      97
15      92
1       80
3       40
5       31
19      13
4       13
Name: theme, dtype: int64

посмотрела сколько анеков какой темы получилось в контрольной выборке

In [53]:
df.to_csv("anek.csv", index=False)

проверю точность модели, а то что-то мне кажется...

In [54]:
X = cvec.transform(themed_df['parsed_aneks']).astype(int)
y = themed_df['theme']

In [55]:
from sklearn.svm import LinearSVC
model = LinearSVC(C=0.01)

In [56]:
from sklearn.model_selection import train_test_split
X_tr, X_te, y_tr, y_te = train_test_split(X, y)

тут я работаю на всей обучающей выборке, разделяя ее на обучающую и контрольную

In [57]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import accuracy_score

использую классификатор KNeighborsClassifier

In [59]:
model.fit(X_tr, y_tr)

LinearSVC(C=0.01)

In [60]:
preds_tr = model.predict(X_tr)
preds_te = model.predict(X_te)
print('для обучающей: ' + str(accuracy_score(y_tr, preds_tr)))
print('для контрольной: ' + str(accuracy_score(y_te, preds_te)))

для обучающей: 0.7529307675293077
для контрольной: 0.5340557275541795


0.534 не очень хорошо, но времени мало... можно было бы конечно убрать некоторые очень редкие классы (и уменьшить частые), или регуляризовать - будет время сделаю...

а теперь попробую что-нибудь погенерировать...

In [61]:
themed_df.to_csv("anek.csv", index=False)

In [64]:
anekdoti = pd.read_csv('anek.csv')

In [249]:
import markovify
import re
import spacy

In [261]:
list_of_themes = ['', 'vova', 'adults', 'computers', 'blondes', 'piter', 'black', 'kinder', 'gabrovska', 'medical', 'school', 'alcohol', 'political', 'auto', 'child', 'russians', 'jews', 'stirlitz', 'different', 'sport']
def make_text_of_all_aneks_themed(i, t):
    about_smth = anekdoti[anekdoti["theme"] == i]
    smth = '\n\n'.join(about_smth['anek'].tolist())
    smth = smth.replace('\r', '')
    with open(f"{theme}.txt", "w", encoding='utf-8') as file:
        file.write(smth)

def make_new_themed_anek(t):
    with open(f"{t}.txt", encoding='utf-8') as f:
        text = f.read()
    text_model = markovify.Text(text)
    
    generated_anek = []
    
    for i in range(2):
        a = text_model.make_sentence()
        if a != None:
            generated_anek.append(str(a))
    
    text = '\n'.join(generated_anek)
    
    return text

а тут я решила не делать все одно и то же как раньше)), а сделать функцию! которая записывает все анеки 1 темы в текстовый файл и 2 функцию для генерации нового анекдота по теме (разделение на темы нужно, чтобы не получалась полная абстракция) на основе цепей маркова

In [262]:
for index, theme in enumerate(list_of_themes):
    if index == 0:
        continue
    make_text_of_all_aneks_themed(index, list_of_themes[index])

сделала кучу ткст файлов с анекдотами

In [263]:
inp = 'piter'
text = make_new_themed_anek(inp)
print(str(text))

Петька в смокинге с букетом и говорит: - Пойдём в дом, покажу.
Уже в дверях Василий Иванович достает Анку расспросами: - Анка, а y вас тpyсы в гоpошек...


тут я смотрела работает ли код

попробуем сделать бота который будет выдавать рандомный анек

In [264]:
import telebot
bot = telebot.TeleBot("5248946987:AAEgWI5LUojBD6S5nqcn7VwklC5WuzPtyFU", parse_mode=None)

dict_users_states = {}

создала бота

In [265]:
keyboard1 = telebot.types.ReplyKeyboardMarkup()
keyboard1.add("/help", "привет", "пока", "рандомный анек про вовочку", 'рандомный анек 18+', 'рандомный анек про компьютеры', 'рандомный анек про блондинок', 'рандомный анек про петю', 'черный юмор', 'про детей', 'габровский анек', 'про медиков', 'про школьников и студентов', 'про алкоголь', 'политический анек', 'про автомобили', 'для детей', 'про русских', 'про евреев', 'про штирлица', 'разное', 'про спорт', 'скинь анек поностальгировать')

создала меню

In [266]:
@bot.message_handler(commands=["start"])
def send_welcome(message):
    bot.reply_to(message, "Привет!/start", reply_markup=keyboard1)

In [267]:
@bot.message_handler(commands=["help"])
def bot_messages(message):
    text = '''этот бот рассказывает сгенерированный нейросетью анекдот по выбранной теме'''
    bot.send_message(message.chat.id, text)

In [268]:
@bot.message_handler(content_types=["text"])
def send_text(message):
    if dict_users_states.get(message.chat.id) != "stop_now_word":
        if message.text.lower() == "привет":
            bot.send_message(message.chat.id, "— Привет, как в кино сходила? У Маринки как дела? Кошка нашлась? Ребенок не болеет? Жалко, что вы с Ваней расстались, он нормальный был.\n— А ты вообще кто?\n— А я статусы твои читаю…\n")
        elif message.text.lower() == "пока":
            bot.send_message(message.chat.id, "- Здгавствуйте...\n- До свидания!")
        elif message.text.lower() == 'рандомный анек про вовочку':
            inp = 'vova'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'рандомный анек 18+':
            inp = 'adults'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'рандомный анек про компьютеры':
            inp = 'computers'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'рандомный анек про блондинок':
            inp = 'blondes'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'рандомный анек про петю':
            inp = 'piter'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'черный юмор':
            inp = 'black'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про детей':
            inp = 'kinder'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'габровский анек':
            inp = 'gabrovska'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про медиков':
            inp = 'medical'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про школьников и студентов':
            inp = 'school'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про алкоголь':
            inp = 'alcohol'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'политический анек':
            inp = 'political'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про автомобили':
            inp = 'auto'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'для детей':
            inp = 'child'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про русских':
            inp = 'russians'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про евреев':
            inp = 'jews'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про штирлица':
            inp = 'stirlitz'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'разное':
            inp = 'different'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'про спорт':
            inp = 'sport'
            annekdot = make_new_themed_anek(inp)
            bot.send_message(message.chat.id, annekdot)
        elif message.text.lower() == 'скинь анек поностальгировать':
            bot.send_photo(message.chat.id, photo=open('ded.jpg', "rb"))

сделала "тело" бота

In [269]:
bot.polling()